cmd/kops/rotate_ca.go

```diff
@@ -0,0 +1,310 @@
+		Signer:         fi.CertificateIDCA,
+		Type:           "ca",
+		Subject:        *subjectPkix,
+		AlternateNames: []string{},
+		Serial:         serial,
+	}
+	cert, privateKey, _, err := pki.IssueCert(&req, keyStore)
+	if err != nil {
+		return err
+	}
+	err = keyStore.StoreKeypair(fi.CertificateIDCA, cert, privateKey)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func rotateCAUpdateServiceAccounts(ctx context.Context, cluster *kops.Cluster, caBundle []byte) error {
+	klog.Info("updating ServiceAccounts with a new CA bundle")
+
+	caBundle64 := base64.StdEncoding.EncodeToString(caBundle)
+	caBundle64Bytes := []byte(caBundle64)
+
+	contextName := cluster.ObjectMeta.Name
+	configLoadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	config, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		configLoadingRules,
+		&clientcmd.ConfigOverrides{CurrentContext: contextName}).ClientConfig()
+	if err != nil {
+		return fmt.Errorf("cannot load kubecfg settings for %q: %v", contextName, err)
+	}
+
+	k8sClient, err := kubernetes.NewForConfig(config)
+
+	if err != nil {
+		return fmt.Errorf("cannot build kubernetes api client for %q: %v", contextName, err)
+	}
+
+	secretClient := k8sClient.CoreV1().Secrets("")
+
+	secrets, err := secretClient.List(ctx, v1.ListOptions{
+		FieldSelector: "type=kubernetes.io/service-account-token",
+	})
+	if err != nil {
+		return err
+	}
+
+	for _, secret := range secrets.Items {
+		secret.Data["ca.crt"] = caBundle64Bytes
+		secretClient.Update(ctx, &secret, v1.UpdateOptions{})
+	}
+
+	return nil
+}
+
+func rotateCAUpdateCluster(ctx context.Context, cluster *kops.Cluster, pool *fi.CertificatePool, f *util.Factory, out io.Writer, exportAdmin bool) error {
+	caBundle, err := pool.AsBytes()
+	if err != nil {
+		return fmt.Errorf("failed to encode ca bundle: %w", err)
+	}
+
+	//Update service accounts to trust old and new CA
+	err = rotateCAUpdateServiceAccounts(ctx, cluster, caBundle)
+	if err != nil {
+		return fmt.Errorf("error updating ServiceAccounts: %v", err)
+	}
+
+	adminTTL, _ := time.ParseDuration("0")
+	if exportAdmin {
+		adminTTL, _ = time.ParseDuration("18h")
+	}
+
+	//New kubeconfig with bundled CA so we trust both new and old api servers
+	RunExportKubecfg(ctx, f, out, &ExportKubecfgOptions{admin: adminTTL}, []string{})
+
+	klog.Info("rotating all nodes")
+
+	//Update nodes first. This will make kubelet trust new and old CA.
+	ruo := &RollingUpdateOptions{}
+	ruo.InitDefaults()
+	ruo.Yes = true
+	ruo.ClusterName = cluster.ObjectMeta.Name
+	ruo.Force = true
+	ruo.InstanceGroupRoles = []string{"node"}
+
+	err = RunRollingUpdateCluster(ctx, f, out, ruo)
+	if err != nil {
+		return fmt.Errorf("failed to rotate cluster: %v", err)
+	}
+
+	klog.Info("rotating the control plane")
+
+	//Update masters. This will issue new certs for k8s using the new CA.
+	//New nodes, service accounts etc will use new CA
+	ruo.InstanceGroupRoles = []string{"master", "apiserver"}
+
```

Comment:
While the new certs/keys currently take effect without an `apply_cluster`, this is undesirable and there are plans to move more of this stuff to userdata and/or versioned files in the store. We should ensure that changing either the set of trusted CAs or the active CA causes the control plane nodes to be treated as "NeedsUpdate". We should also allow for a future need to do an `apply_cluster`.